w11h03/test/pgdp/networking/SocketTest.java

```diff
@@ -0,0 +1,441 @@
+        client.getOutputStream().write(SERVER_HELLO);
+        client.getOutputStream().flush();
+        getAllBytes(connectThread, client);
+
+        Thread clientThread = new Thread(() -> {
+            dataHandler.sendMessage(message);
+        });
+        // Send Server Hello
+        clientThread.start();
+        Thread.sleep(SOCKET_TIMEOUT);
+        // Write ACK
+        var buffer = getAllBytes(clientThread, client);
+        assertEquals(0x01, buffer[0], "Incorrect first byte");
+        var actualLength = copyOf(buffer, 1, 3, "Message length");
+
+        // See below
+        OrBuilder
+                .assertThat(() -> assertArrayEquals(new byte[]{0x01, 0x23}, actualLength))
+                .or(() -> assertArrayEquals(new byte[]{0x01, 0x0D}, actualLength))
+                .withMessage("Incorrect message length")
+                .run();
+
+        String actualMessage = new String(buffer, StandardCharsets.UTF_8).substring(3);
+        int messageLength = actualMessage.length();
+
+        // THIS REMOVES NULL-BYTES AT THE END OF THE STRING AND DELIBERATELY ALLOWS FOR A FAULTY REPRESENTATION OF THE MESSAGE
+        // This is because the transformation into a byte-array is done incorrectly in the template
+        // https://zulip.in.tum.de/#narrow/stream/1525-PGdP-W11H03/topic/StandardCharsets.2EUTF_8.2Eencode.28message.29.2Earray.28.29/near/909726
+        if (buffer.length == 294) {
+            actualMessage = actualMessage.replace("\00", "");
+        }
+        assertEquals(message, actualMessage, "Incorrect message content.");
+
+        // Assert end of output
+        assertEquals(messageLength + 7, buffer.length, "Transferred bytes did not end after expected message.");
+    }
+
+    @Test
+    @Order(9)
+    @DisplayName("sendMessage() – huge")
+    void testMessagesHuge() throws IOException, InterruptedException {
+        // Ideally, we would use resources here, but that seems difficult with the current tests setup
+        String lipsum = Files.readString(Path.of("./test/pgdp/networking/lipsum.txt"));
+        String lipsumTruncated = Files.readString(Path.of("./test/pgdp/networking/lipsum_truncated.txt"));
+        Thread connectThread = new Thread(this::connect);
+        connectThread.start();
+        Socket client = server.accept();
+        client.getOutputStream().write(SERVER_HELLO);
+        client.getOutputStream().flush();
+        getAllBytes(connectThread, client);
+
+        Thread clientThread = new Thread(() -> {
+            dataHandler.sendMessage(lipsum);
+        });
+        // Send Server Hello
+        clientThread.start();
+        Thread.sleep(SOCKET_TIMEOUT);
+        // Write ACK
+        var buffer = getAllBytes(clientThread, client);
+        assertEquals(0x01, buffer[0], "First byte");
+        byte[] actualLength = copyOf(buffer, 1, 3, "Message length");
+        assertArrayEquals(new byte[] {(byte) 0xff, (byte) 0xff}, actualLength, "Message length was incorrect.");
```

Comment:
> new byte[] {(byte) 0xff, (byte) 0xff}

Isn't this `{-1, -1}`? And makes this sense?

![image](https://user-images.githubusercontent.com/24459435/213873580-3817f823-69a7-4488-9fdf-d150d79a9c1c.png)

https://www.jdoodle.com/iembed/v0/CA0

Reply:
It is. I just figured that because length is always positive (and encoded in 2 bytes), this representation would make it easier to understand what the tests are expecting.

Reply:
Ah makes sense, my issue was that I forgot to use the 2nd and 3rd byte for the length of the message 😅